data/virt_autotest/guest_unattended_installation_files/opensuse_tumbleweed_kvm_hvm_guest_x86_64.xml

```diff
@@ -174,5 +174,17 @@
       <username>root</username>
     </user>
   </users>
+  <scripts>
+    <!-- permit root login, password login and pubkeys login -->
+    <init-scripts config:type="list">
+      <script>
+        <source><![CDATA[
+echo -e "PermitRootLogin yes\nPubkeyAuthentication yes\nPasswordAuthentication yes\nAuthorizedKeysFile      .ssh/authorized_keys" > /etc/ssh/sshd_config
+[ -d /root/.ssh ] || mkdir /root/.ssh; echo "##Authorized-Keys##" >> /root/.ssh/authorized_keys
```

Comment:
Can it be made more tolerable -- no matter how these settings be changed in future(on or off, partially or all), the script can set up the required configurations correctly?  Same as below.

Reply:
sshd_config has been removed in TW. this shell script is the easiest way to create a new sshd_config over autoyast. We may have more cautious way to check if the file exist then blabla.., like 

if sshd_config exists
  sed /xxx/xxx/s if xxx exists
else
  echo ...

autoyast debugging is more complex than simple shell environment ---- have to run an entire autoyast installation each time and there is no detail debug information. I'll have a try, If it goes well, I will improve it. Otherwise, I'd better keep the simplest method to save effort :-).

Reply:
if removed sshd_config file as expected in TW, would be better to add comments in opensuse_tumbleweed_XX.xml file to let us clear this changed.

Reply:
@Julie-CAO suse's sshd is just openssh deamon. According to openssh sshd document:
https://man.openbsd.org/sshd.8

-f config_file
Specifies the name of the configuration file. The default is /etc/ssh/sshd_config. sshd refuses to start if there is no configuration file.

So if sshd config is not available, then there is no chance that sshd  can be started, let alone "whether root login is permitted". So I think the root cause might be sshd config is missing instead of root login is not allowed.

Reply:
I checked sshd.service is running without errors when TW is installed.

root login is really not permitted, please refer to the bug in my comment: bsc#1173067, the progress ticket from other team https://progress.opensuse.org/issues/93829, and the PR(https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/12697) which says 'not permit root login' will be adapted by SLE16 and Leap 16.

Reply:
This bug mentions that something found in config file, but the config file now is missing from TW I think. So how can I know what default settings are ? I think there might be a product issue if default config file is not available for sshd.

Reply:
@alice-suse do you know where the upstream is adopted in TW? where or how can I know the TW we tested is supposed to be? I did not find any document for TW, I referenced the docs for openSUSE Leap.

Reply:
Talked with @rfan1 offline that the removal of sshd_config is the fact for TW and it is expected for SLE16 and Leap 16, the info is coming from developer, no Jira feature or document about it. Anyway, sshd is not our test target but a necessary procedure for virtualization testing, so I prefer to let it work rather than investigate whether it is reasonable.

Reply:
It is acceptable at the moment I think.

Reply:
> @alice-suse do you know where the upstream is adopted in TW? where or how can I know the TW we tested is supposed to be? I did not find any document for TW, I referenced the docs for openSUSE Leap.

@Julie-CAO You can get official product changes(mean to have changes, rather than random) according to below quote from https://confluence.suse.com/pages/viewpage.action?spaceKey=qasleapac2&title=Factory-first+discussion+meeting+20210527. 
```
TW product change notice: there can be several ways to follow:
1) check in https://planet.opensuse.org/en/ the  posts with title "openSUSE Tumbleweed – Review of the week YYYY/WW", eg http://dominique.leuenberger.net/blog/2021/05/opensuse-tumbleweed-review-of-the-week-2021-21/
2) https://download.opensuse.org/tumbleweed/iso/ files with name Changes.[6 digit date].txt  
3) check "News portals" and "Mailing lists" in https://en.opensuse.org/openSUSE:Communication_channels
4) https://build.opensuse.org/project/show/openSUSE:Factory package changelogs

Note: confirmed with Dominique  Leuenberger, 1) and 2) can  give a rather complete insight of what happened and also a glimpse at what is coming.
```

Reply:
> if removed sshd_config file as expected in TW, would be better to add comments in opensuse_tumbleweed_XX.xml file to let us clear this changed.

more comments added.

Reply:
> Can it be made more tolerable -- no matter how these settings be changed in future(on or off, partially or all), the script can set up the required configurations correctly? Same as below.

changed the shell script in autoyast in following cases:
case1: "#key=value" => "key=value"
case2: key does not exist => add the key on the end of sshd_config
case3: sshd_config does not exist => create and add all the keys

Reply:
find the change about sshd in TW changelogs:
'''OpenSSH: move of the distro default config to /usr/etc. Admin config is supposed to be in /etc. Make sure to check the .rpmsave files after the update. Best to put config snippets in /etc/ssh/sshd_config.d/*.conf files to inherit distro config changes'''
'''- Change vendor configuration dir from /usr/share/ssh/ to
  /usr/etc/ssh/.
- Remove upgrade enablement hack. This has been fixed in
  systemd-rpm-macros (bsc#1180083).
- Add support for vendor provided configuration files in
  /usr/share/ssh/ (openssh-8.4p1-vendordir.patch)
- Move configuration files from /etc/ssh/ to /usr/share/ssh/
- Drop openssh-7.7p1-allow_root_password_login.patch to prevent login
  as root via password by default (is also upstream default). Comment
  indicates that this was a temporary meassure that we now had for
  five years, time to get rid of it (bsc#1173067)'''

tried /etc/ssh/sshd_config.d/sshd_config does not take effect, so still keep it /etc/ssh/sshd_config.